presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.execution.buffer;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockEncodingSerde;
+import io.airlift.compress.Compressor;
+import io.airlift.compress.Decompressor;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+
+import static com.facebook.presto.block.BlockSerdeUtil.readBlock;
+import static com.facebook.presto.block.BlockSerdeUtil.writeBlock;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.compress.lz4.Lz4RawCompressor.maxCompressedLength;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class PagesSerde
+{
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final Optional<Compressor> compressor;
+    private final Optional<Decompressor> decompressor;
+
+    public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> compressor, Optional<Decompressor> decompressor)
+    {
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.compressor = requireNonNull(compressor, "compressor is null");
+        this.decompressor = requireNonNull(decompressor, "decompressor is null");
+        checkArgument(compressor.isPresent() == decompressor.isPresent(), "compressor and decompressor must both be present");
```

Comment:
Haha I thought about this for a while - you will only see this failure if one of them is present, so the error message works well. In terms of reading the code, it's definitely confusing though :P